cpp/src/parquet/types.h

```diff
@@ -64,8 +64,12 @@ struct Type {
     DOUBLE = 5,
     BYTE_ARRAY = 6,
     FIXED_LEN_BYTE_ARRAY = 7,
+
+    // This parquet type does not actually exist (AFAIK) and is used to
+    // create proper type traits
+    LARGE_BYTE_ARRAY = 8,
```

Comment:
I don't think this is right and the parquet internal physical type should be left unchanged.

We should always derive the target arrow binary type from the arrow reader/writer properties.

Reply:
Thanks for the feedback. Let me see if I understand it correctly: the parquet layer should remain unchanged, while the arrow layer should be able to convert / construct a `LargeBinaryArray` from a `parquet::ByteArrayType`. In order to do that, you are suggesting that the only changes are a new `arrow::LargeBinaryBuilder` and a runtime selection between LARGE and non-large binary builder inside `arrow::ChunkedRecordReader`?

Reply:
Yes, I agree that `TypedRecordReader` would be a challenge. However, we know the actual builder type from the reader properties so we can create correct builder accordingly. One option would be adding a `ByteArrayEncodingTraits` which is similar to `EncodingTraits`. Or we can extend `EncodingTraits` with an extra parameter to indicate int32 or int64 for binary types. Both approaches are ugly though.

Reply:
Yes... I am experimenting both, but they do look hackish..

Reply:
Btw, any other suggestions or a preference here? I am not sure what to do, everything I have tried seems to break the current design.. @wgtmac

Reply:
As `EncodingTraits` resides in the public header `parquet/encoding.h` which may already be used externally, we'd better not break compatibility. What about this?

```cpp
template <typename T, bool EnableLargeBinary = false>
struct EncodingTraits;

...

template <>
struct EncodingTraits<ByteArrayType, /*EnableLargeBinary=*/false> {
  using Encoder = ByteArrayEncoder;
  using Decoder = ByteArrayDecoder;
  struct Accumulator {
    std::unique_ptr<::arrow::BinaryBuilder> builder;
    std::vector<std::shared_ptr<::arrow::Array>> chunks;
  };
  using ArrowType = ::arrow::BinaryType;
  using DictAccumulator = ::arrow::Dictionary32Builder<::arrow::BinaryType>;
};

template <>
struct EncodingTraits<ByteArrayType, /*EnableLargeBinary=*/true> {
  using Encoder = ByteArrayEncoder;
  using Decoder = ByteArrayDecoder;
  struct Accumulator {
    std::unique_ptr<::arrow::LargeBinaryBuilder> builder;
    std::vector<std::shared_ptr<::arrow::Array>> chunks;
  };
  using ArrowType = ::arrow::LargeBinaryType;
  using DictAccumulator = ::arrow::Dictionary32Builder<::arrow::LargeBinaryBuilder>;
};
```

Reply:
This was the first thing I have tried, but this would require changing a lot of classes like: `TypedRecordReader`, `TypedColumnReaderImpl`, `ColumnReaderImplBase` and adapt a few methods along the way. Would you accept something like this?

Reply:
Actually, a lot of other classes need to be modified because of that..

Reply:
So I agree with @wgtmac here, the large type is a concept of Arrow and shouldn't be intermingled into parquet types.  I need to look more closely at the code to have a suggestion here.

Reply:
I also agree with you guys, that does seem to be the right thing. And iirc, it'll eliminate a lot of code, making it simpler. The challenge is to do this without breaking current design

Reply:
@emkornfield @wgtmac Did you guys get a chance to look into this?

Reply:
Now I see the problem comes from here:
```cpp
template <typename DType>
class TypedDecoder : virtual public Decoder {
 public:
  using T = typename DType::c_type;

  virtual int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
                          int64_t valid_bits_offset,
                          typename EncodingTraits<DType>::Accumulator* out) = 0;

  virtual int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
                          int64_t valid_bits_offset,
                          typename EncodingTraits<DType>::DictAccumulator* builder) = 0;
};
```

So, I still prefer extending `EncodingTraits` with a new parameter like below
```cpp
template <typename T, bool EnableLargeBinary = false>
struct EncodingTraits;
```

To workaround issues with the decoder, we can apply similar extension:
```cpp
template <typename DType, bool EnableLargeBinary = false>
class TypedDecoder : virtual public Decoder {};

template <typename DType, bool EnableLargeBinary = false>
class DictDecoder : virtual public TypedDecoder<DType, EnableLargeBinary> {};
```

WDYT?

Reply:
Did you try to actually implement it? I might be missing something, but it seems like tons of other changes are needed, starting wtih `TypedRecordReader`.

Reply:
Yes, those readers and decoders templated with <typename DType> are probably needed to be changed the same way. @pitrou @emkornfield @mapleFU WDYT?

Reply:
Sorry I may have selected the wrong diff. Yes, you current proposal looks better to me.